beetsplug/metasync/itunes.py

```diff
@@ -0,0 +1,97 @@
+# This file is part of beets.
+# Copyright 2015, Tom Jaspers.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""Synchronize information from iTunes's library
+"""
+from contextlib import contextmanager
+import os
+import shutil
+import tempfile
+import plistlib
+from time import mktime
+
+from beets import util
+from beets.dbcore import types
+from beets.library import DateType
+from beets.util.confit import ConfigValueError
+from beetsplug.metasync import MetaSource
+
+
+@contextmanager
+def create_temporary_copy(path):
+    temp_dir = tempfile.mkdtemp()
+    temp_path = os.path.join(temp_dir, 'temp_itunes_lib')
+    shutil.copyfile(path, temp_path)
+    try:
+        yield temp_path
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+class Itunes(MetaSource):
+
+    item_types = {
+        'itunes_rating':      types.INTEGER,  # 0..100 scale
+        'itunes_playcount':   types.INTEGER,
+        'itunes_skipcount':   types.INTEGER,
+        'itunes_lastplayed':  DateType(),
+        'itunes_lastskipped': DateType(),
+    }
+
+    def __init__(self, config, log):
+        super(Itunes, self).__init__(config, log)
+
+        # Load the iTunes library, which has to be the .xml one (not the .itl)
+        library_path = util.normpath(config['itunes']['library'].get(str))
+
+        try:
+            self._log.debug(
+                u'loading iTunes library from {0}'.format(library_path))
+            with create_temporary_copy(library_path) as library_copy:
+                raw_library = plistlib.readPlist(library_copy)
+        except IOError as e:
+            raise ConfigValueError(u'invalid iTunes library: ' + e.strerror)
+        except Exception:
+            # It's likely the user configured their '.itl' library (<> xml)
+            if os.path.splitext(library_path)[1].lower() != '.xml':
+                hint = u': please ensure that the configured path' \
+                       u' points to the .XML library'
+            else:
+                hint = ''
+            raise ConfigValueError(u'invalid iTunes library' + hint)
+
+        # Convert the library in to something we can query more easily
+        self.collection = {
+            (track['Name'], track['Album'], track['Album Artist']): track
+            for track in raw_library['Tracks'].values()}
```

Comment:
That was my original intention (and how I would prefer) it, but iTunes saves path in Unix style on all systems. This is conflicting with the way paths are saved in beets for Windows.

In *Itunes Library.xml*:
`file://localhost/G:/Experiments/Alt-J/An%20Awesome%20Wave/01%20Intro.mp3`

And in *beets*:
`G:\Experiments\Alt-J\An Awesome Wave\01 Intro.mp3`

I'm gonna try to resort to converting the iTunes path to match beets' style. Will take a while to test it properly on Windows system though, but using a path is probably better in the long run.